tests/atomic_fence/atomic_fence.cpp

```diff
@@ -0,0 +1,293 @@
+  static constexpr test_type TestType = TestT::value;
+
+ public:
+  void operator()(const std::string& memory_order_name,
+                  const std::string& memory_scope_name,
+                  const std::string& test_type_name) {
+    SECTION(sycl_cts::section_name(std::string("Check atomic_fence with "
+                                               "memory_order = ") +
+                                   memory_order_name +
+                                   " and scope = " + memory_scope_name +
+                                   " and test_type = " + test_type_name)
+                .create()) {
+      auto queue = sycl_cts::util::get_cts_object::queue();
+      if (!check_memory_order_scope_capabilities(queue, MemoryOrder,
+                                                 MemoryScope, memory_order_name,
+                                                 memory_scope_name)) {
+        return;
+      }
+      sycl::memory_order order_write = MemoryOrder;
+      sycl::memory_order order_read = MemoryOrder;
+      if (sycl::memory_order::release == order_read) {
+        order_read = sycl::memory_order::acquire;
+      }
+
+      // Count of retries in the check cycle
+      constexpr size_t RETRY_COUNT = 256;
+
+      bool res = true;
+      int sync = 0;
+      int data = 0;
+      int value = expected_val;
+
+      // These global_range and local_range values provide a check in one group
+      // when test_type = single_group, and between four groups when
+      // test_type = between_groups
+      sycl::range<1> global_range(2);
+      if (test_type::between_groups == TestType) {
+        global_range = sycl::range<1>(8);
+      }
+      sycl::range<1> local_range(2);
+
+      {
+        sycl::buffer<bool> res_buf(&res, sycl::range<1>(1));
+        sycl::buffer<int> sync_buffer(&sync, sycl::range<1>(1));
+        sycl::buffer<int> data_buffer(&data, sycl::range<1>(1));
+        queue.submit([&](sycl::handler& cgh) {
+          auto res_acc =
+              res_buf.template get_access<sycl::access_mode::write>(cgh);
+          auto sync_flag_acc = get_accessor(cgh, sync_buffer);
+          auto data_acc = get_accessor(cgh, data_buffer);
+          cgh.parallel_for(
+              sycl::nd_range<1>(global_range, local_range),
+              [=](sycl::nd_item<1> nditem) {
+                auto g = nditem.get_group();
+                sycl::atomic_ref<int, sycl::memory_order::relaxed,
+                                 sycl::memory_scope::work_group>
+                    sync_flag(sync_flag_acc[0]);
+                int* data = &data_acc[0];
+                if ((test_type::single_group == TestType && g.leader()) ||
+                    (test_type::between_groups == TestType &&
+                     0 == nditem.get_global_linear_id())) {
+                  *data = value;
+                  sycl::atomic_fence(order_write, MemoryScope);
+                  sync_flag = 1;
+                } else {
```

Comment:
Yes, all around this subtle kernel you should add some comments explaining what is happening, the hypothesis, etc.

Reply:
Done